File: plugin/src/main/groovy/com/github/gfx/ribbonizer/plugin/RibbonizerExtension.java
package com.github.gfx.ribbonizer.plugin;

import com.android.build.gradle.api.ApplicationVariant;
import com.github.gfx.ribbonizer.CustomColorRibbonBuilder;
import com.github.gfx.ribbonizer.FilterBuilder;
import com.github.gfx.ribbonizer.GrayScaleBuilder;
import com.github.gfx.ribbonizer.GrayRibbonBuilder;
import com.github.gfx.ribbonizer.GreenRibbonBuilder;
import com.github.gfx.ribbonizer.YellowRibbonBuilder;
import com.github.gfx.ribbonizer.filter.ColorRibbonFilter;
import com.github.gfx.ribbonizer.filter.GrayScaleFilter;

import java.awt.image.BufferedImage;
import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.function.Consumer;

public class RibbonizerExtension {

    public static String NAME = "ribbonizer";

    Set<String> forcedVariantsNames = new HashSet<>();

    Set<String> iconNames = new HashSet<>();

    List<FilterBuilder> filterBuilders = new ArrayList<>();

    public RibbonizerExtension() {
    }

    public Set<String> getIconNames() {
        return iconNames;
    }

    /**
     * @param resNames Names of icons. For example "@drawable/ic_launcher", "@mipmap/icon"
     */
    public void setIconNames(Collection<String> resNames) {
        iconNames = new HashSet<>(resNames);
    }

    /**
     * @param resNames Names of icons. For example "@drawable/ic_launcher", "@mipmap/icon"
     */
    public void iconNames(Collection<String> resNames) {
        setIconNames(resNames);
    }

    /**
     * @param resNames Names of icons. For example "@drawable/ic_launcher", "@mipmap/icon"
     */
    public void iconNames(String... resNames) {
        setIconNames(Arrays.asList(resNames));
    }

    /**
     * @param resName A name of icons. For example "@drawable/ic_launcher", "@mipmap/icon"
     */
    public void iconName(String resName) {
        iconNames.add(resName);
    }

    public Set<String> getForcedVariantsNames() {
        return forcedVariantsNames;
    }

    public void forcedVariantsNames(String... variantsNames) {
        forcedVariantsNames = new HashSet<>(Arrays.asList(variantsNames));
    }

    public List<FilterBuilder> getFilterBuilders() {
        return filterBuilders;
    }

    public void setFilterBuilders(Collection<FilterBuilder> filterBuilders) {
        this.filterBuilders = new ArrayList<>(filterBuilders);
    }

    public void builder(FilterBuilder filterBuilder)
            throws IllegalAccessException, InstantiationException {
        this.filterBuilders.clear();
        this.filterBuilders.add(filterBuilder);
    }

    // utilities

    public GrayScaleFilter grayScaleFilter(ApplicationVariant variant, File iconFile) {
        return (GrayScaleFilter) new GrayScaleBuilder().apply(variant, iconFile);
    }

    public ColorRibbonFilter grayRibbonFilter(ApplicationVariant variant, File iconFile) {
        return (ColorRibbonFilter) new GrayRibbonBuilder().apply(variant, iconFile);
    }

    public ColorRibbonFilter yellowRibbonFilter(ApplicationVariant variant, File iconFile) {
        return (ColorRibbonFilter) new YellowRibbonBuilder().apply(variant, iconFile);
    }

    public ColorRibbonFilter greenRibbonFilter(ApplicationVariant variant, File iconFile) {
        return (ColorRibbonFilter) new GreenRibbonBuilder().apply(variant, iconFile);
    }

    public ColorRibbonFilter customColorRibbonFilter(ApplicationVariant variant, File iconFile, String nm) {
        return (ColorRibbonFilter) new CustomColorRibbonBuilder(nm).apply(variant, iconFile);
    }
}

File: ribbonizer-plugin/src/main/kotlin/com/shogo82148/ribbonizer/plugin/RibbonizerExtension.kt
package com.shogo82148.ribbonizer.plugin

import com.android.build.gradle.api.ApplicationVariant
import com.shogo82148.ribbonizer.*
import com.shogo82148.ribbonizer.filter.ColorRibbonFilter
import com.shogo82148.ribbonizer.filter.GrayScaleFilter
import com.shogo82148.ribbonizer.resource.Resource
import com.shogo82148.ribbonizer.resource.Variant
import org.gradle.internal.serialization.Transient.Var
import java.io.File
import java.util.function.Consumer

@Suppress("unused") // The fields of the Variant are used from Gradle DSL.
open class RibbonizerExtension {
    private var _iconNames: MutableSet<String> = HashSet()

    val iconNames: Set<String>
        get() = _iconNames

    fun setIconNames(resNames: Collection<String>) {
        _iconNames = HashSet(resNames)
    }

    fun iconNames(resNames: Collection<String>) {
        _iconNames = HashSet(resNames)
    }

    fun iconNames(vararg resNames: String) {
        _iconNames = HashSet(listOf(*resNames))
    }

    fun iconName(resName: String) {
        _iconNames.add(resName)
    }

    private var _forcedVariantsNames: MutableSet<String> = HashSet()

    val forcedVariantsNames: Set<String>
        get() = _forcedVariantsNames

    fun setForcedVariantsNames(variantNames: Collection<String>) {
        _forcedVariantsNames = HashSet(variantNames)
    }

    fun forcedVariantsNames(variantNames: Collection<String>) {
        _forcedVariantsNames = HashSet(variantNames)
    }

    fun forcedVariantsNames(vararg variantNames: String) {
        _forcedVariantsNames = HashSet(listOf(*variantNames))
    }

    private var _filterBuilders: MutableList<FilterBuilder> = ArrayList()

    val filterBuilders: List<FilterBuilder>
        get() = _filterBuilders

    fun setFilterBuilders(filterBuilders: Collection<FilterBuilder>) {
        _filterBuilders = ArrayList(filterBuilders)
    }

    fun builder(filterBuilder: (Variant, File) -> Consumer<Resource>?) {
        val fb = object: FilterBuilder{
            override fun apply(variant: Variant, file: File): Consumer<Resource>? {
                return filterBuilder(variant, file)
            }
        }
        _filterBuilders.clear()
        _filterBuilders.add(fb)
    }

    // utilities

    fun grayScaleFilter(variant: Variant, iconFile: File): GrayScaleFilter {
        return GrayScaleBuilder().apply(variant, iconFile) as GrayScaleFilter
    }

    fun grayRibbonFilter(variant: Variant, iconFile: File): ColorRibbonFilter {
        return GrayRibbonBuilder().apply(variant, iconFile) as ColorRibbonFilter
    }

    fun yellowRibbonFilter(variant: Variant, iconFile: File): ColorRibbonFilter {
        return YellowRibbonBuilder().apply(variant, iconFile) as ColorRibbonFilter
    }

    fun greenRibbonFilter(variant: Variant, iconFile: File): ColorRibbonFilter {
        return GreenRibbonBuilder().apply(variant, iconFile) as ColorRibbonFilter
    }

    fun customColorRibbonFilter(
        variant: Variant,
        iconFile: File,
        nm: String
    ): ColorRibbonFilter {
        return CustomColorRibbonBuilder(nm).apply(variant, iconFile) as ColorRibbonFilter
    }

    companion object {
        const val NAME = "ribbonizer"
    }
}